ow_power_system/src/PredictionHandler.cpp

```diff
@@ -0,0 +1,145 @@
+// The Notices and Disclaimers for Ocean Worlds Autonomy Testbed for Exploration
+// Research and Simulation can be found in README.md in the root directory of
+// this repository.
+
+// See PredictionHandler.h for a summary of the purpose of this file.
+
+#include <chrono>
+#include <ros/package.h>
+#include <ros/ros.h>
+#include <sensor_msgs/JointState.h>
+#include <PrognoserFactory.h>
+
+#include "PredictionHandler.h"
+
+using namespace PCOE;
+
+// The indices use to access temperature information.
+// This might change to median SOC or RUL index or fixed percentile.
+// (predecessor comment to Summer 2022, unsure exactly what it means)
+static constexpr int BATTERY_TEMPERATURE_INDEX = 0;
+static constexpr int MODEL_TEMPERATURE_INDEX = 1;
+
+/**
+ * Constructs a new prediction handler that subscribes to battery EoD
+ * predictions for the specified source and on the specified message
+ * bus.
+ **/
+PredictionHandler::PredictionHandler(MessageBus& bus, const std::string& src) :
+                                     m_bus(bus)
+{
+  m_bus.subscribe(this, src, MessageId::BatteryEod);
+  m_event_ready = false;
+}
+
+/**
+ * Unsubscribes the prediction handler from the message bus.
+ **/
+PredictionHandler::~PredictionHandler()
+{
+  m_bus.unsubscribe(this);
+}
+
+/**
+ * The message bus will call this function each time the predictor publishes
+ * a new battery EoD prediction.
+ **/
+void PredictionHandler::processMessage(const std::shared_ptr<Message>& message)
+{
+  using namespace std::chrono;
+  // The prediction handler only ever subscribes to the BatteryEoD message
+  // id, which should always be a ProgEventMessage, so this should always
+  // succeed.
+  auto prediction_msg = dynamic_cast<ProgEventMessage*>(message.get());
```

Comment:
The use of `dynamic_cast` (which seems best avoided in general) combined with `auto` is a bit confusing.  Could we either eliminate the cast or declare `prediction_msg` as `ProgEventMessage*` ?

Reply:
This was part of the example taken from GSAP's async prognoser; I wouldn't be familiar with the ``Message`` pointer type & converting it to ``ProgEventMessage``. I moved the declaration from ``auto`` to ``ProgEventMessage*``, but I wouldn't be sure how to modify this line to move away from ``dynamic_cast``.

I'm curious: what is the reason ``dynamic_cast`` should be avoided? I never learned about it in university.